rfcs/2021-02-04-5802-event-metadata.md

```diff
@@ -0,0 +1,206 @@
+the metadata will need to be filled in) and the "into" methods (`fn
+into_log` and `fn into_metric`). Implementations for `Deref` and
+`DerefMut` will simplify porting existing code to the new structure.
+
+### Vector Protocol
+
+The event protocol is used for both the transmission of events between
+Vector instances and the storage of events in disk buffers. This
+metadata is only relevant within a single instance of Vector, so should
+not be transmitted across the wire. However, for the latter, the
+protobuf definition will need to be expanded to include the metadata by
+adding the appropriate structures.
+
+```protobuf
+message EventWithMetadata {
+  EventWrapper data = 1;
+  EventMetadata metadata = 2;
+}
+
+message EventMetadata {
+  google.protobuf.Timestamp first_timestamp = 1;
+  repeated SourceEvent source_events = 2;
+  EventDeliveryStatus status = 3;
+}
+
+message SourceEvent {
+  google.protobuf.Timestamp timestamp = 1;
+  string source_name = 2;
+  string event_id = 3;
+}
+
+enum EventDeliveryStatus {
+  …TBD…
+}
+```
+
+### Visibility
+
+This metadata will not be visible to users, either through remap or JSON
+transforms, unless a use case can demonstrate the need for it. Given the
+above structure of the metadata, there are no user-modifiable parts.
+
+## Doc-level Proposal
+
+N/A
+
+## Rationale
+
+This addition is required to support the two features mentioned in the
+motivation. These provide users with:
+
+1. metrics about how long events are taking to reach their destination,
+   which can be used in meta-analysis about the observation framework
+   itself, and
+2. assurance that events are fully retained between when they are
+   generated and when they are stored and processed, but reducing the
+   number of ways they may be lost in transit.
+
+Once in place, this metadata may also enable additional features not yet
+envisioned. For example, the VRL compiler collects type information on
+fields, but this can be expensive for arrays. Transforms such as
+`add_fields` could build a field type cache to improve the performance
+of this feature.
+
+The introduction of `SmallVec` in the metadata structure is driven by
+the assumption that most events will have a single source throughout
+their lifetime. `SmallVec` is a data structure that can inline a (fixed)
+number of elements before allocating memory. By using this feature, this
+common case can avoid an extra allocation and improve data locality.
+
+## Prior Art
+
+N/A
+
+## Drawbacks
+
+- The event data is a critical component of Vector. Any changes to this
+  structure impacts virtually the entire system.
+- All transforms that combine or duplicate events will have extra
+  complications in handing the metadata. This particularly applies to
+  the Lua and WASM scripting transforms.
+- This metadata will increase the memory required by Vector for any
+  given event flow.
+- Recording, copying, and finalizing the metadata will increase the CPU
+  usage.
```

Comment:
I know this violates the "make it work and then make it fast" rule, but I think it's worth discussing performance more before we accept this RFC. Especially since sufficient performance will be required before we can ship this feature.

There have been talks about using data structures, or patterns, that "share" data, avoiding the need to allocate entirely new metadata for every event. For example, if the `source_name` is the same for every event coming from that source, we do not need to allocate a string for every event. This might be obvious, but I think it's worth elucidating on this more and possibly benchmarking ideas.

Reply:
I think this is a benefit of the struct-based approach described here. As we add each field we can optimize the representation for that specific use case (e.g. `Arc<String>` for source name). That's opposed to a map-based implementation where we'd need to consider all future uses.